core/src/main/java/org/apache/accumulo/core/util/compaction/CompactionServicesConfig.java

```diff
@@ -0,0 +1,186 @@
+          .anyMatch(prop -> prop == null || aconf.isPropertySet(prop, true));
+
+      if (defaultServicePropsSet) {
+
+        String warning = String.format(
+            "The deprecated property %s was set. Properties with the prefix %s "
+                + "were also set, which replace the deprecated properties. The deprecated "
+                + "property was therefore ignored.",
+            Property.TSERV_MAJC_MAXCONCURRENT.getKey(), defaultServicePrefix);
+
+        deprecationWarningConsumer.accept(warning);
+
+      } else {
+        String numThreads = aconf.get(Property.TSERV_MAJC_MAXCONCURRENT);
+
+        // Its possible a user has configured the other compaction services, but not the default
+        // service. In this case want to produce a config with the default service configs
+        // overridden using deprecated configs.
+
+        HashMap<String,String> configsCopy = new HashMap<>(configs);
+
+        Map<String,String> defaultServiceConfigs =
+            Map.of(defaultServicePrefix + "planner", DefaultCompactionPlanner.class.getName(),
+                defaultServicePrefix + "planner.opts.executors",
+                "[{'name':'deprecated', 'numThreads':" + numThreads + "}]");
+
+        configsCopy.putAll(defaultServiceConfigs);
+
+        String warning = String.format(
+            "The deprecated property %s was set. Properties with the prefix %s "
+                + "were not set, these should replace the deprecated properties. The old "
+                + "properties were automatically mapped to the new properties in process "
+                + "creating : %s.",
+            Property.TSERV_MAJC_MAXCONCURRENT.getKey(), defaultServicePrefix,
+            defaultServiceConfigs);
+
+        deprecationWarningConsumer.accept(warning);
+
+        configs = Map.copyOf(configsCopy);
+      }
+    }
+
+    return configs;
+
+  }
+
+  public CompactionServicesConfig(AccumuloConfiguration aconf,
+      Consumer<String> deprecationWarningConsumer) {
+    this.deprecationWarningConsumer = deprecationWarningConsumer;
+    Map<String,String> configs = getConfiguration(aconf);
+
+    configs.forEach((prop, val) -> {
+
+      var suffix = prop.substring(Property.TSERV_COMPACTION_SERVICE_PREFIX.getKey().length());
+      String[] tokens = suffix.split("\\.");
+      if (tokens.length == 4 && tokens[1].equals("planner") && tokens[2].equals("opts")) {
+        getOptions().computeIfAbsent(tokens[0], k -> new HashMap<>()).put(tokens[3], val);
+      } else if (tokens.length == 2 && tokens[1].equals("planner")) {
+        getPlanners().put(tokens[0], val);
+      } else if (tokens.length == 3 && tokens[1].equals("rate") && tokens[2].equals("limit")) {
+        var eprop = Property.getPropertyByKey(prop);
+        if (eprop == null || aconf.isPropertySet(eprop, true)
+            || !isDeprecatedThroughputSet(aconf)) {
+          getRateLimits().put(tokens[0], ConfigurationTypeHelper.getFixedMemoryAsBytes(val));
```

Comment:
Should access the private maps (planners, rateLimits and options) directly here instead of the public methods (getOptions(), getplanners() and getRateLimits()). The methods are public so could be overridden and could change the behavior of the constructor. I don't think that is what you want since this code seems very specific to that one property.

Reply:
Addressed in 76b4e70